modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/InlineUtils.java

```diff
@@ -0,0 +1,147 @@
+                return Math.min(maxVarlenInlineSize, ((VarlenNativeType) nativeType).length());
+
+            case DECIMAL:
+                // It can be converted to float for greater or lesser comparisons.
+                return Float.BYTES;
+
+            case NUMBER:
+                // It can be converted to int for greater or lesser comparisons.
+                return Integer.BYTES;
+
+            default:
+                throw new IllegalArgumentException("Unknown type " + spec);
+        }
+    }
+
+    /**
+     * Calculates inline size for {@link BinaryTuple}, given its format.
+     *
+     * @param indexDescriptor Index descriptor.
+     * @param maxVarlenInlineSize Maximum inline size of variable length column in bytes.
+     * @return Inline size in bytes, not more than the {@link #MAX_BINARY_TUPLE_INLINE_SIZE}.
+     */
+    static int binaryTupleInlineSize(IndexDescriptor indexDescriptor, int maxVarlenInlineSize) {
+        List<? extends ColumnDescriptor> columns = indexDescriptor.columns();
+
+        boolean hasNullColumns = columns.stream().anyMatch(ColumnDescriptor::nullable);
+
+        // 2-byte size class covers most cases.
+        int sizeClass = 2;
+
+        if (columns.stream().allMatch(c -> c.type().spec().fixedLength())) {
+            int totalFixedColumnSize = columns.stream().mapToInt(c -> c.type().sizeInBytes()).sum();
+
+            sizeClass = Math.min(sizeClass, BinaryTupleCommon.flagsToEntrySize(BinaryTupleCommon.valueSizeToFlags(totalFixedColumnSize)));
+        }
+
+        int inlineSize = BinaryTupleCommon.HEADER_SIZE
+                + (hasNullColumns ? BinaryTupleCommon.nullMapSize(columns.size()) : 0)
+                + columns.size() * sizeClass;
+
+        for (int i = 0; i < columns.size() && inlineSize < MAX_BINARY_TUPLE_INLINE_SIZE; i++) {
+            inlineSize += inlineSize(columns.get(i).type(), maxVarlenInlineSize);
+        }
+
+        return Math.min(inlineSize, MAX_BINARY_TUPLE_INLINE_SIZE);
+    }
+
+    /**
+     * Calculates the inline size for the {@link BinaryTuple} that will be stored in the index {@link BplusInnerIo} and {@link BplusLeafIo}.
+     * Minimum number of items in a {@link BplusInnerIo} is {@code 2}, so that there is no significant performance drop.
+     *
+     * @param pageSize Page size in bytes.
+     * @param indexBplusIoHeaderSize Index {@link BplusInnerIo} and {@link BplusLeafIo} header size in bytes.
+     * @param indexDescriptor Index descriptor.
+     * @param maxVarlenInlineSize Maximum inline size of variable length column in bytes.
+     * @return Inline size in bytes, not more than the {@link #MAX_BINARY_TUPLE_INLINE_SIZE}.
+     */
+    public static int binaryTupleBplusIoInlineSize(
+            int pageSize,
+            int indexBplusIoHeaderSize,
+            IndexDescriptor indexDescriptor,
+            int maxVarlenInlineSize
+    ) {
+        int innerNodePayloadSize = pageSize - BplusInnerIo.HEADER_SIZE - PARTITIONLESS_LINK_SIZE_BYTES;
```

Comment:
I think we should introduce a constant like `CHILD_LINK_SIZE` in BplusInnerIo, what do you think? Having explicit mentions of partitionless links is too mouthful

Reply:
Add it